### ISSUE TITLE
[onert] Support mnist model training

### ISSUE THREAD

**Issue by jyoungyun** [Samsung/ONE#11389]
## Target model

MNIST CNN model from [kaggle](https://www.kaggle.com/code/amyjang/tensorflow-mnist-cnn-tutorial)

### Model summary

```
Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #
=================================================================
 conv2d (Conv2D)             (None, 28, 28, 32)        832

 conv2d_1 (Conv2D)           (None, 28, 28, 32)        25632

 max_pooling2d (MaxPooling2  (None, 14, 14, 32)        0
 D)

 dropout (Dropout)           (None, 14, 14, 32)        0

 conv2d_2 (Conv2D)           (None, 14, 14, 64)        18496

 conv2d_3 (Conv2D)           (None, 14, 14, 64)        36928

 max_pooling2d_1 (MaxPoolin  (None, 7, 7, 64)          0
 g2D)

 dropout_1 (Dropout)         (None, 7, 7, 64)          0

 flatten (Flatten)           (None, 3136)              0

 dense (Dense)               (None, 128)               401536

 dropout_2 (Dropout)         (None, 128)               0

 dense_1 (Dense)             (None, 10)                1290

=================================================================
Total params: 484714 (1.85 MB)
Trainable params: 484714 (1.85 MB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________
```

### Model structure

<img src="https://github.com/Samsung/ONE/assets/7223627/a84e1496-9184-4d7b-aca4-fab29aaa244a" height=300/>

### Train

- total epoch : 5
- batch size : 64
- loss function : CategoricalCrossentropy
- optimizer : RMSProp
   - epsilon : 1e-08

### File

**tflite**

**circle**

<details>
<summary>code</summary>

model.py
```python
#!/usr/bin/env python3

import os
import argparse
import tensorflow as tf
import numpy as np

def normalize_img(image, label):
  """Normalizes images: `uint8` -> `float32`."""
  return tf.cast(image, tf.float32) / 255., label

def initParser():
    parser = argparse.ArgumentParser()
    parser.add_argument('-o', '--outdir', required=True, help='output dir path')

    return parser.parse_args()

def createModel():
    model = tf.keras.models.Sequential([
        tf.keras.layers.Conv2D(32, (5,5), padding='same', activation='relu', input_shape=(28, 28, 1)),
        tf.keras.layers.Conv2D(32, (5,5), padding='same', activation='relu'),
        tf.keras.layers.MaxPool2D(),
        tf.keras.layers.Dropout(0.25),
        tf.keras.layers.Conv2D(64, (3,3), padding='same', activation='relu'),
        tf.keras.layers.Conv2D(64, (3,3), padding='same', activation='relu'),
        tf.keras.layers.MaxPool2D(strides=(2,2)),
        tf.keras.layers.Dropout(0.25),
        tf.keras.layers.Flatten(),
        tf.keras.layers.Dense(128, activation='relu'),
        tf.keras.layers.Dropout(0.5),
        tf.keras.layers.Dense(10, activation='softmax')
    ])

    model.summary()
    return model

def trainModel(model):
    mnist = tf.keras.datasets.mnist
    (x_train, y_train), (x_test, y_test) = mnist.load_data()

    x_train=x_train.reshape(x_train.shape[0], x_train.shape[1], x_train.shape[2], 1)
    x_train=x_train / 255.0
    x_test = x_test.reshape(x_test.shape[0], x_test.shape[1], x_test.shape[2], 1)
    x_test=x_test/255.0

    y_train = tf.one_hot(y_train.astype(np.int32), depth=10)
    y_test = tf.one_hot(y_test.astype(np.int32), depth=10)

    model.compile(
        optimizer=tf.keras.optimizers.RMSprop(epsilon=1e-08),
        loss='categorical_crossentropy',
        metrics=['acc']
    )
    
    model.fit(
        x_train, y_train,
        batch_size=64,
        epochs=5,
        validation_split=0.1,
    )

def saveModel(model, pbPath):
    tf.keras.models.save_model(model, pbPath)

def saveTfliteModel(pbPath):
    converter = tf.lite.TFLiteConverter.from_saved_model(pbPath)
    tflite_model = converter.convert()

    tflite_path = os.path.join(pbPath, 'model.tflite')
    with open(tflite_path, 'wb') as f:
          f.write(tflite_model)

if __name__ == "__main__":

    args = initParser()

    model = createModel()

    trainModel(model)

    saveModel(model, args.outdir)

    saveTfliteModel(args.outdir)
```

tensorflow_run.py
```
#!/usr/bin/env python3

import argparse
import tensorflow as tf
import numpy as np

def initParse():
    parser = argparse.ArgumentParser()
    parser.add_argument('-m', '--model', required=True, help='model filename')
    parser.add_argument('-i', '--image', required=True, help='image filename')
    parser.add_argument('-l', '--label', required=True, help='label filename')

    return parser.parse_args()

def train(modelPath, imagePath, labelPath):
    np_images = np.fromfile(imagePath, dtype=[('data', np.float32)])
    np_labels = np.fromfile(labelPath, dtype=[('data', np.float32)])
    
    train_images = tf.reshape(tf.convert_to_tensor(np_images, tf.float32), (1,28,28))
    train_labels = tf.reshape(tf.convert_to_tensor(np_labels, tf.int32), (1,10))
    
    print(train_images[0])
    print(train_labels[0])
    
    model = tf.keras.models.load_model(modelPath)

    model.summary()

    model.compile(
        optimizer=tf.keras.optimizers.SGD(learning_rate=0.01),
        loss=tf.keras.losses.MeanSquaredError(reduction=tf.keras.losses.Reduction.SUM_OVER_BATCH_SIZE),
        metrics=[tf.keras.metrics.MeanSquaredError()],
    )

    model.fit(
      train_images,
      train_labels,
      epochs=5,
      batch_size=1,
      shuffle=False,
#      steps_per_epoch=None,
#      max_queue_size=1,
#      validation_split=0.0,
#      initial_epoch=0
    )
    
if __name__ == "__main__":

    args = initParse()

    train(args.model, args.image, args.label)

```

mnist_kaggle.cfg
```
[onecc]
one-import-tflite=True
one-optimize=True

[one-import-tflite]
input_path=model.tflite
output_path=mnist_kaggle.circle

[one-optimize]
input_path=mnist_kaggle.circle
output_path=mnist_kaggle.opt.circle
```
</details>

**Comment by jyoungyun** [comment]
### Todo

#### Train operations

- [ ] MaxPool2D
- [ ] SoftMax

#### Optimizer

- [ ] RMSProp

#### Loss

- [ ] Categorical Cross entropy

**Comment by jyoungyun** [comment]
### Check

- [x] FullyConnected
   - Same loss, meaningful latency
   - Verified from Keras mnist sample model
- [x] Reshape
   - Same loss, meaningful latency
   - Verified from Keras mnist sample model
- [x] Conv2d
   - Same loss
   - https://github.sec.samsung.net/jy910-yun/Personal/issues/71#issuecomment-2339244
- [x] MaxPool2d
   - Same loss
   - https://github.sec.samsung.net/jy910-yun/Personal/issues/71#issuecomment-2338699
- [x] Softmax
   - Same loss, meaningful latency
   - https://github.sec.samsung.net/jy910-yun/Personal/issues/122#issuecomment-2338165

**Comment by jyoungyun** [comment]
### Test Full model w/o Dropout

data_length: 1000
batch_size: 1000
loss: mse
optimizer: sgd

The loss value matches.

tensorflow
```
$ python3 tensorflow_run.py -m ./mnist_kaggle/mnist_kaggle_nodropout/ -i ../data/fashion_mnist/train.input.1000.bin -l ../data/fashion_mnist/train.output.1000.bin --data_length 1000 --epoch 5 --batch_size 1000
Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #
=================================================================
 conv2d (Conv2D)             (None, 28, 28, 32)        832

 conv2d_1 (Conv2D)           (None, 28, 28, 32)        25632

 max_pooling2d (MaxPooling2  (None, 14, 14, 32)        0
 D)

 conv2d_2 (Conv2D)           (None, 14, 14, 64)        18496

 conv2d_3 (Conv2D)           (None, 14, 14, 64)        36928

 max_pooling2d_1 (MaxPoolin  (None, 7, 7, 64)          0
 g2D)

 flatten (Flatten)           (None, 3136)              0

 dense (Dense)               (None, 128)               401536

 dense_1 (Dense)             (None, 10)                1290

=================================================================
Total params: 484714 (1.85 MB)
Trainable params: 484714 (1.85 MB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________
Epoch 1/5
1/1 [==============================] - 0s 481ms/step - loss: 0.1546 - mean_squared_error: 0.1546
Epoch 2/5
1/1 [==============================] - 0s 249ms/step - loss: 0.1543 - mean_squared_error: 0.1543
Epoch 3/5
1/1 [==============================] - 0s 247ms/step - loss: 0.1540 - mean_squared_error: 0.1540
Epoch 4/5
1/1 [==============================] - 0s 240ms/step - loss: 0.1537 - mean_squared_error: 0.1537
Epoch 5/5
1/1 [==============================] - 0s 243ms/step - loss: 0.1533 - mean_squared_error: 0.1533
==========================
Total time: 20.1516
```

onert_train
```
$ ./Product/out/bin/onert_train --modelfile ../Personal/models/mnist/mnist_kaggle/mnist_kaggle_nodropout/mnist_kaggle.circle --load_input:raw ../Personal/models/data/fashion_mnist/train.input.1000.bin --load_expected:raw ../Personal/models/data/fashion_mnist/
train.output.1000.bin --learning_rate 0.01 --loss 0 --data_length 1000 --optimizer 0 --epoch 5 --batch_size 1000
Model Expected Filename ../Personal/models/data/fashion_mnist/train.output.1000.bin
Model Input Filename ../Personal/models/data/fashion_mnist/train.input.1000.bin
Model Filename ../Personal/models/mnist/mnist_kaggle/mnist_kaggle_nodropout/mnist_kaggle.circle
Epoch 1/5 - time: 1197.706ms/step - loss: [0] 0.1546
Epoch 2/5 - time: 1195.347ms/step - loss: [0] 0.1543
Epoch 3/5 - time: 1194.028ms/step - loss: [0] 0.1540
Epoch 4/5 - time: 1186.045ms/step - loss: [0] 0.1537
Epoch 5/5 - time: 1187.833ms/step - loss: [0] 0.1533
===================================
MODEL_LOAD   takes 0.8670 ms
PREPARE      takes 309.6970 ms
EXECUTE      takes 5963.6030 ms
- Epoch 1      takes 1197.7060 ms
- Epoch 2      takes 1195.3470 ms
- Epoch 3      takes 1194.0280 ms
- Epoch 4      takes 1186.0450 ms
- Epoch 5      takes 1187.8330 ms
===================================
```

**Comment by jyoungyun** [comment]
### Test Full model w/o Dropout

data_length: 1000
batch_size: 1000
loss: categorical crossentropy
optimizer: adam

tensorflow
```
$ python3 tensorflow_run.py -m ./mnist_kaggle/mnist_kaggle_nodropout/ -i ../data/fashion_mnist/train.input.1000.bin -l ../data/fashion_mnist/train.output.1000.bin --data_length 1000 --epoch 5 --loss cce --optimizer adam  --batch_size 1000 --learning_rat 0.001 --loss_reduction_type sum_over_batch_size
2023-11-30 21:00:18.504804: I tensorflow/core/util/port.cc:110] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2023-11-30 21:00:18.564785: I tensorflow/tsl/cuda/cudart_stub.cc:28] Could not find cuda drivers on your machine, GPU will not be used.
2023-11-30 21:00:19.229949: I tensorflow/tsl/cuda/cudart_stub.cc:28] Could not find cuda drivers on your machine, GPU will not be used.
2023-11-30 21:00:19.272457: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 AVX_VNNI FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
2023-11-30 21:00:22.428172: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT
Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #
=================================================================
 conv2d (Conv2D)             (None, 28, 28, 32)        832

 conv2d_1 (Conv2D)           (None, 28, 28, 32)        25632

 max_pooling2d (MaxPooling2  (None, 14, 14, 32)        0
 D)

 conv2d_2 (Conv2D)           (None, 14, 14, 64)        18496

 conv2d_3 (Conv2D)           (None, 14, 14, 64)        36928

 max_pooling2d_1 (MaxPoolin  (None, 7, 7, 64)          0
 g2D)

 flatten (Flatten)           (None, 3136)              0

 dense (Dense)               (None, 128)               401536

 dense_1 (Dense)             (None, 10)                1290

=================================================================
Total params: 484714 (1.85 MB)
Trainable params: 484714 (1.85 MB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________
Epoch 1/5
1/1 [==============================] - 2s 2s/step - loss: 8.5378
Epoch 2/5
1/1 [==============================] - 1s 599ms/step - loss: 5.5028
Epoch 3/5
1/1 [==============================] - 0s 397ms/step - loss: 3.7889
Epoch 4/5
1/1 [==============================] - 0s 303ms/step - loss: 2.8428
Epoch 5/5
1/1 [==============================] - 0s 301ms/step - loss: 2.3365
==========================
Total time: 21.9340
```

onert_train
```
$ ./Product/out/bin/onert_train --modelfile ../Personal/models/mnist/mnist_kaggle/mnist_kaggle_nodropout/mnist_kaggle.circle --load_input:raw ../Personal/models/data/fashion_mnist/train.input.1000.bin --load_expected:raw ../Personal/models/data/fashion_mnist/train.output.1000.bin --loss 1 --optimizer 1 --epoch 5 --batch_size 1000 --learning_rate 0.001 --loss_reduction_type 1
Model Expected Filename ../Personal/models/data/fashion_mnist/train.output.1000.bin
Model Input Filename ../Personal/models/data/fashion_mnist/train.input.1000.bin
Model Filename ../Personal/models/mnist/mnist_kaggle/mnist_kaggle_nodropout/mnist_kaggle.circle
Epoch 1/5 - time: 1312.930ms/step - loss: [0] 8.5378
Epoch 2/5 - time: 1300.508ms/step - loss: [0] 5.5020
Epoch 3/5 - time: 1305.560ms/step - loss: [0] 3.7881
Epoch 4/5 - time: 1345.110ms/step - loss: [0] 2.8421
Epoch 5/5 - time: 1335.561ms/step - loss: [0] 2.3358
===================================
MODEL_LOAD   takes 40.6610 ms
PREPARE      takes 372.2580 ms
EXECUTE      takes 6602.6600 ms
- Epoch 1      takes 1312.9300 ms
- Epoch 2      takes 1300.5080 ms
- Epoch 3      takes 1305.5600 ms
- Epoch 4      takes 1345.1100 ms
- Epoch 5      takes 1335.5610 ms
===================================
```

**Comment by jyoungyun** [comment]
Done!
